storm-server/src/test/java/org/apache/storm/scheduler/resource/strategies/scheduling/TestLargeCluster.java

```diff
@@ -0,0 +1,458 @@
+                }
+            }
+            for (int rack = 14 ; rack < 16 ; rack++) {
+                double cpu = 1200; // %percent
+                double mem = 42_480; // MB
+                for (int superInRack = 0; superInRack < numSupersPerRack ; superInRack++) {
+                    createAndAddOneSupervisor(rack, superInRack, cpu - 100 * (superInRack % 2), mem, numPorts, retList);
+                }
+            }
+            return retList;
+        }
+    }
+
+    /**
+     * Create a large cluster, read topologies and configuration from resource directory and schedule.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testLargeCluster() throws Exception {
+        boolean uniformSupervisors = false; // false means non-uniform supervisor distribution
+
+        Map<String, SupervisorDetails> supervisors = createSupervisors(uniformSupervisors);
+
+        TopologyDetails[] topoDetailsArray = createTopoDetailsArray(false);
+        Assert.assertTrue("No topologies found", topoDetailsArray.length > 0);
+        Topologies topologies = new Topologies(topoDetailsArray);
+
+        Config confWithDefaultStrategy = new Config();
+        confWithDefaultStrategy.putAll(topoDetailsArray[0].getConf());
+        confWithDefaultStrategy.put(Config.TOPOLOGY_SCHEDULER_STRATEGY, DefaultResourceAwareStrategy.class.getName());
+
+        INimbus iNimbus = new INimbusTest();
+        Cluster cluster = new Cluster(iNimbus, new ResourceMetrics(new StormMetricsRegistry()), supervisors, new HashMap<>(),
+                topologies, confWithDefaultStrategy);
+
+        scheduler = new ResourceAwareScheduler();
+
+        List<Class> classesToDebug = Arrays.asList(DefaultResourceAwareStrategy.class,
+                GenericResourceAwareStrategy.class, ResourceAwareScheduler.class,
+                Cluster.class // count calls to calculateSharedOffHeapNodeMemory()
+        );
+        Level logLevel = Level.INFO ; // switch to Level.DEBUG for verbose otherwise Level.INFO
+        classesToDebug.forEach(x -> Configurator.setLevel(x.getName(), logLevel));
+        long startTime = System.currentTimeMillis();
+        scheduler.prepare(confWithDefaultStrategy, new StormMetricsRegistry());
+        scheduler.schedule(topologies, cluster);
+        long endTime = System.currentTimeMillis();
+        LOG.info("Scheduling Time: {} topologies in {} seconds", topoDetailsArray.length, (endTime - startTime) / 1000.0);
+
+        for (TopologyDetails td : topoDetailsArray) {
+            TestUtilsForResourceAwareScheduler.assertTopologiesFullyScheduled(cluster, td.getName());
+        }
+
+        // Remove topology and reschedule it
+        java.util.Random random = new java.util.Random();
+        for (int i = 0 ; i < topoDetailsArray.length ; i++) {
+            startTime = System.currentTimeMillis();
+            int iTopo = i; // random.nextInt(topoDetailsArray.length);
```

Comment:
Which one we want to keep? ` iTopo = i` or `random.nextInt(topoDetailsArray.length);`

Reply:
i for now.

Reply:
Maybe clean it up to avoid confusion. Pick one

Reply:
changed